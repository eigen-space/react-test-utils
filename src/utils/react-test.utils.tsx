import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import 'jest-styled-components';
import { AnyDictionary } from '@eigenspace/common-types';
import { EmulateActionType } from '../enums/emulate.action.enum';
import { BoxWrapper, ReactTestUtilsConfig } from '..';

export class ReactTestUtils {
    private static themeWrapper?: React.ComponentClass;

    static init(config: ReactTestUtilsConfig): void {
        ReactTestUtils.themeWrapper = config.themeWrapper;
    }

    static renderComponent<P = {}, S = {}>(
        component: React.ReactElement<P>,
        props?: P
    ): ReactWrapper<P, S, React.Component> {
        if (props) {
            return mount<React.Component, P, S>(ReactTestUtils.themeWrap(React.cloneElement(component, props)));
        }

        return mount<React.Component, P, S>(ReactTestUtils.themeWrap(component));
    }

    static defaultPropsWrapper(func: Function): () => void {
        return jest.fn(() => func());
    }

    /**
     * Used to wrap component into box to test absolute positioned elements.
     */
    static boxWrap(component: React.ReactElement, box: BoxWrapper, props?: AnyDictionary): JSX.Element {
        const style: BoxWrapper = { ...box };
        style.width = style.width ? `${style.width}px` : 'fit-content';
        style.height = style.height ? `${style.height}px` : 'auto';

        const childComponent = React.cloneElement(component, props);
        return <div style={style}> {childComponent}</div>;
    }

    static themeWrap(component: React.ReactElement): React.ReactElement {
        if (!ReactTestUtils.themeWrapper) {
            return component;
        }

        return <ReactTestUtils.themeWrapper>{component}</ReactTestUtils.themeWrapper>;
    }

    /**
     * This method add class to element to be able use puppeteer's actions like hover and focus.
     */
    static emulateAction(wrapper: ReactWrapper, action: EmulateActionType, payload = ''): void {
        // @ts-ignore
        wrapper.getDOMNode().setAttribute(action, payload);
    }

    static waitWhileAsyncPollIsNotEmpty(): Promise<void> {
        return new Promise(resolve => setImmediate(resolve));
    }
}
