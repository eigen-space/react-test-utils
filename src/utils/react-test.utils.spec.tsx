import * as React from 'react';
import { ReactTestUtils } from './react-test.utils';

describe('ReactTestUtils', () => {

    describe('#boxWrap', () => {
        const component = <input/>;

        it('should set defined box value', async () => {
            const box = { width: '150', height: '50' };
            const wrapper = ReactTestUtils.boxWrap(component, box);
            await expect({ component: wrapper }).toMatchComponentImageAsync();
        });

        it('should set default value, any key in box is undefined', async () => {
            const wrapper = ReactTestUtils.boxWrap(component, {});
            await expect({ component: wrapper }).toMatchComponentImageAsync();
        });
    });
});
