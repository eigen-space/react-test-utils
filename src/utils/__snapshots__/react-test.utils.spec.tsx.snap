// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactTestUtils #boxWrap should set default value, any key in box is undefined 1`] = `"oZxr5m+RPfYdCkrtaIWzPv0icK0="`;

exports[`ReactTestUtils #boxWrap should set defined box value 1`] = `"fjWvti/KglhzmNmQ5/JNodoctz8="`;
